Add blog post about binding an external service (#5828)
redhat-developer/odo: docs/website/blog/2022-06-14-binding-external-service.md

---
title: Binding an external service with odo v3
author: Philippe Martin
author_url: https://github.com/feloy
author_image_url: https://github.com/feloy.png
tags: ["binding"]
slug: binding-external-service-with-odo-v3
---

How to bind an external service using odo v3

<!--truncate-->

When developers are working on a micro-service that needs to access a database or another service, 
they may want to provide to their application the address and the necessary credentials to access
this service as simply as possible.

In this article, we are saying that we are *binding* the service to the application.

Using the Service Binding Operator and creating some Kubernetes resources for
each service you want to be bindable, you can make the life easier for developers.

## Creating a Service resource to redirect to the external service

To expose an external service from inside a Kubernetes cluster, you can create a *Headless* Service,
and manually create the Enpoints to access this external service.

> Here an example, to connect to an external Redis service on IP 192.168.1.10 and port 6379:

Reviewer: ```suggestion
Here is an example, to connect to an external Redis service on IP 192.168.1.10 and port 6379:
```

Also, a more realistic example would be using an external DNS name, no? Maybe you could use an example with an external domain name instead, or mention this as well..

```
kind: Service
apiVersion: v1
metadata:
  name: redis
  namespace: external-services
spec:
  type: ClusterIP
  ports:
  - port: 6379
    targetPort: 6379

---

kind: Endpoints
apiVersion: v1
metadata:
  name: redis
  namespace: external-services
subsets:
- addresses:
  - ip: 192.168.1.10
  ports:
  - port: 6379
```

Note that we have created these resources in a `external-services` namespace, which is a dedicated namespace
to store external services information, accessible by all developers.

You can find more information about creating Service resources to access external services [here](https://docs.openshift.com/dedicated/3/dev_guide/integrating_external_services.html) or [here](https://www.youtube.com/watch?v=fvpq4jqtuZ8).

## Storing the credentials into a Secret resource

The Redis instance is protected by a password, and you may want to store this password into a Secret resource,
so it can be used by applications.

> The developers may want to *mount* this Secret into their application's Pod, but Secret are mountable only

Reviewer: ```suggestion
The developers may want to *mount* this Secret into their application's Pod, but Secrets are mountable only
```

```

indicates to the SBO to inject a `host` variable into the application, with a value computed based on the
`metadata.name` and `metadata.namespace` of the Service resource. In this example, the value `redis.external-services.svc.cluster.local`
will be given to the `host` variable.

The second annotation

```
service.binding: path={.metadata.name}-credentials,objectType=Secret
```

indicates to the SBO to inject the values defined in the Secret, whose name is the name of the Service resource
followed by `-credentials` (`redis-credentials` in our example), into the application. In this example, the variable `password` 
with a value `MyEasyPassword` will be injected into the applications's Pod.

## Adding a ServiceBinding to the Devfile

To define a ServiceBinding, we need information (group, version, kind, name and namespace) about the Application and the Service.

In our example, the service is a Kubernetes Service (group "", version "v1" and kind "Service") named `redis`
in the `external-services` namespace, 

The application will be the Deployment resource (group "apps", version "v1", kind "Deployment") created by odo when you run `odo dev`.

> By convention, the Deployment name will be the name of the Devfile (in the `.metadata.name` field) followed by `-app` (`my-nodejs-app-app` in our example).

Reviewer: You are mentioning `my-nodejs-app-app`, but I didn't see it mentioned before.

Author: Fixed

> You don't have to specify the namespace, as the Deployment will be in the same namespace as the Deployment.

Reviewer: I don't understand this part of your sentence (`Deployment will be in the same namespace as the Deployment`) :p

Author: fixed

The option `bindAsFiles` indicates to the SBO to create files into the Pod's container, each file having the name 
of a credential, and containing the value of the credential.

```
apiVersion: binding.operators.coreos.com/v1alpha1
kind: ServiceBinding
metadata:
  name: binding-to-redis
spec:
  application:
    group: apps
    version: v1
    kind: Deployment
    name: my-nodejs-app-app
  services:
  - group: ""
    version: v1
    kind: Service
    name: redis
    namespace: external-services
  bindAsFiles: true
```

You can create a file `kubernetes/redis.yaml` in your directory containing this snippet,
and add a Kubernetes component into your Devfile referring to this YAML file:

```
components:

- name: binding-to-redis
  kubernetes:
    uri: kubernetes/redis.yaml
```

By adding this Kubernetes component to your Devfile, when you run `odo dev`, the ServiceBinding resource defined
in the `kubernetes/redis.yaml` file will be created into the cluster, and the Service Binding Operator will inject
into the application's Pod the `host` and `password` necessary to connect to the Redis external service.

> 

Reviewer: A suggestion: maybe you can end this article by showing `odo dev` in action, and also how `odo describe binding` can help show more details about the binding. To me, this would give a quick hint at how the injected credentials could be consumed by the application developer.

Author: I've added some content to explain how to use the bindings. 
Thanks

```
metadata:
  name: my-nodejs-app
[...]
components:
[...]
- name: binding-to-redis
  kubernetes:
    uri: kubernetes/redis.yaml
```

By adding this Kubernetes component to your Devfile, when you run `odo dev`, the ServiceBinding resource defined
in the `kubernetes/redis.yaml` file will be created into the cluster, and the Service Binding Operator will inject
into the application's Pod the `host` and `password` necessary to connect to the Redis external service.

## Using the variables into the application's code

The Devfile is now ready, and the developer can start accessing the external service from the code. 

The first step to know how the credentials are exposed into the application's container is to start the `odo dev` 
command and to execute the `odo describe binding` command.

Running `odo dev`, you can see that the ServiceBinding resource is deployed to the cluster.

```
$ odo dev
[...]
↪ Deploying to the cluster in developer mode
 ✓  Creating kind ServiceBinding [60ms]
 ✓  Waiting for Kubernetes resources [10s]
 ✓  Syncing files into the container [740ms]
 ✓  Building your application in container on cluster [4s]
 ✓  Executing the application [1s]
[...]
```

> From another terminal, running `odo describe binding` shows you the status of the ServiceBinding:

Reviewer 2: ```suggestion
From another terminal, running `odo describe binding` will show you the status of the ServiceBinding:
```
